// eve-swagger/lib/api/search.js
'use strict';

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CallableInstance = require('../internal/callable-instance');

function defaultSearch(agent, categories, strict, text) {
  return agent.noAuth.get('/v1/search/', {
    query: {
      'categories': categories,
      'search': text,
      'strict': strict
    }
  }).then(function (result) {
    if (categories.length == 1) {
      // Return array for that category
      return result[categories[0]];
    } else {
      // Return everything
      return result;
    }
  });
}

function characterSearch(agent, categories, strict, character, token, text) {
  return agent.auth(token).get('/v2/characters/{character_id}/search/', {
    path: { 'character_id': character },
    query: {
      'categories': categories,
      'search': text,
      'strict': strict
    }
  }).then(function (result) {
    if (categories.length == 1) {
      // Return array for that category
      return result[categories[0]];
    } else {
      // Return everything
      return result;
    }
  });
}

/**
 * An api adapter over the end points handling search and character search via
 * functions in the [search](https://esi.tech.ccp.is/latest/#/Search) ESI
 * endpoints. You should not usually instantiate this directly as its
 * constructor requires an internal api instance.
 *
 * This is a function class so instances of `Search` are functions and can be
 * invoked directly, besides accessing its members. Its default function action
 * is equivalent to {@link Search#get get}.
 */

var Search = function (_CallableInstance) {
  (0, _inherits3.default)(Search, _CallableInstance);

  /**
   * Create a new Search api wrapper with the given configuration. If
   * `categories` is provided then the search is restricted to those
   * categories,
   * otherwise it will search over all categories. If `categories` is a single
   * item, the return result of the search is simplified to be the array of ids
   * for that category (instead of an outer object with keys per category).
   *
   * Categories must from the following enumeration:
   *
   * + `agent`
   * + `alliance`
   * + `character`
   * + `constellation`
   * + `corporation`
   * + `faction`
   * + `inventorytype`
   * + `region`
   * + `solarsystem`
   * + `station`
   * + `wormhole`
   *
   * If `characterId` and `accessToken` are provided then the search will use
   * character specific search end point, and the `structure` category can also
   * be used.
   *
   * @param agent {ESIAgent} The ESI agent
   * @param categories {Array.<String>} Optional; the categories search to
   *     through
   * @param characterId {Number} Optional; the character id of the search
   * @param accessToken {String} Optional; SSO token for the provided
   *     character,
   *    must be provided if `characterId` is given
   */
  function Search(agent) {
    var categories = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
    var characterId = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;
    var accessToken = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : '';
    (0, _classCallCheck3.default)(this, Search);

    var _this = (0, _possibleConstructorReturn3.default)(this, (Search.__proto__ || (0, _getPrototypeOf2.default)(Search)).call(this, function (text) {
      return _this.get(text);
    }));

    _this._agent = agent;
    _this._categories = categories;
    _this._character = characterId;
    _this._token = accessToken;

    if (!categories || categories.length == 0) {
      // Make default full list based on character search (which includes
      // structure)
      _this._categories = ['agent', 'alliance', 'character', 'constellation', 'corporation', 'faction', 'inventorytype', 'region', 'solarsystem', 'station', 'wormhole'];

      if (_this._character != 0 && _this._token != '') {
        _this._categories.push('structure');
      }
    }
    return _this;
  }

  /**
   * @esi_route get_search
   * @esi_param search - text
   * @esi_param !categories
   * @esi_param strict - false
   *
   * However, if the Search instance is configured to search over only a single
   * category then the Promise simply resolves to the array of ids (e.g. the
   * category was fetched from the response dictionary of category to array).
   *
   * @esi_example esi.search('query') strict=false
   *
   * @param {String} text The search terms of the query
   * @return {Promise.<Object>} Or a `Promise.<Array.<Number>>` for a
   *    single-category search.
   */


  (0, _createClass3.default)(Search, [{
    key: 'get',
    value: function get(text) {
      return this._doSearch(text, false);
    }

    /**
     * @esi_route get_search
     * @esi_param search - text
     * @esi_param !categories
     * @esi_param strict - true
     *
     * However, if the Search instance is configured to search over only a single
     * category then the Promise simply resolves to the array of ids (e.g. the
     * category was fetched from the response dictionary of category to array).
     *
     * @esi_example esi.search.strict('query') strict=true
     *
     * @param {String} text The search terms of the query
     * @return {Promise.<Object>} Or a `Promise.<Array.<Number>>` for a
     *    single-category search.
     */

  }, {
    key: 'strict',
    value: function strict(text) {
      return this._doSearch(text, true);
    }
  }, {
    key: '_doSearch',
    value: function _doSearch(text, strict) {
      if (this._character == '' && this._token == '') {
        // Do a default search
        return defaultSearch(this._agent, this._categories, strict, text);
      } else {
        // Do a character search
        return characterSearch(this._agent, this._categories, strict, this._character, this._token, text);
      }
    }
  }]);
  return Search;
}(CallableInstance);

module.exports = Search;
